<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#MIDS---w261-Machine-Learning-At-Scale" data-toc-modified-id="MIDS---w261-Machine-Learning-At-Scale-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>MIDS - w261 Machine Learning At Scale</a></span><ul class="toc-item"><li><span><a href="#Assignment---HW5" data-toc-modified-id="Assignment---HW5-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Assignment - HW5</a></span><ul class="toc-item"><li><span><a href="#INSTRUCTIONS-for-SUBMISSIONS" data-toc-modified-id="INSTRUCTIONS-for-SUBMISSIONS-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>INSTRUCTIONS for SUBMISSIONS</a></span></li></ul></li></ul></li><li><span><a href="#Optional-reading" data-toc-modified-id="Optional-reading-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Optional reading</a></span></li><li><span><a href="#HW-Problems" data-toc-modified-id="HW-Problems-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>HW Problems</a></span></li><li><span><a href="#HW5.0--data-warehouse;-star-schema" data-toc-modified-id="HW5.0--data-warehouse;-star-schema-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>HW5.0  data warehouse; star schema</a></span></li><li><span><a href="#HW5.1-Databases:-3NF;-denormalized" data-toc-modified-id="HW5.1-Databases:-3NF;-denormalized-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>HW5.1 Databases: 3NF; denormalized</a></span></li><li><span><a href="#HW5.2--Memory-backed-map-side" data-toc-modified-id="HW5.2--Memory-backed-map-side-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>HW5.2  Memory-backed map-side</a></span></li><li><span><a href="#HW5.2.1-(OPTIONAL)-Almost-stateless-reducer-side-join" data-toc-modified-id="HW5.2.1-(OPTIONAL)-Almost-stateless-reducer-side-join-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>HW5.2.1 (OPTIONAL) Almost stateless reducer-side join</a></span></li><li><span><a href="#Pairwise-similarity----PHASE-1" data-toc-modified-id="Pairwise-similarity----PHASE-1-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Pairwise similarity  - PHASE 1</a></span><ul class="toc-item"><li><span><a href="#Run-Systems-tests-locally-on-small-datasets-(PHASE1)" data-toc-modified-id="Run-Systems-tests-locally-on-small-datasets-(PHASE1)-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>Run Systems tests locally on small datasets (PHASE1)</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#1:-unit/systems-first-10-lines" data-toc-modified-id="1:-unit/systems-first-10-lines-8.1.0.1"><span class="toc-item-num">8.1.0.1&nbsp;&nbsp;</span>1: unit/systems first-10-lines</a></span></li><li><span><a href="#2:-unit/systems-atlas-boon" data-toc-modified-id="2:-unit/systems-atlas-boon-8.1.0.2"><span class="toc-item-num">8.1.0.2&nbsp;&nbsp;</span>2: unit/systems atlas-boon</a></span></li><li><span><a href="#3:-unit/systems-stripe-docs-test" data-toc-modified-id="3:-unit/systems-stripe-docs-test-8.1.0.3"><span class="toc-item-num">8.1.0.3&nbsp;&nbsp;</span>3: unit/systems stripe-docs-test</a></span></li></ul></li></ul></li><li><span><a href="#5.3-build-stripes-for-all-the-test-data-sets---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="5.3-build-stripes-for-all-the-test-data-sets---run-the-commands-and-insure-that-your-output-matches-the-output-below-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>5.3 build stripes for all the test data sets - run the commands and insure that your output matches the output below</a></span></li><li><span><a href="#5.4-Build-Inverted-Index---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="5.4-Build-Inverted-Index---run-the-commands-and-insure-that-your-output-matches-the-output-below-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>5.4 Build Inverted Index - run the commands and insure that your output matches the output below</a></span><ul class="toc-item"><li><span><a href="#Inverted-Index" data-toc-modified-id="Inverted-Index-8.3.1"><span class="toc-item-num">8.3.1&nbsp;&nbsp;</span>Inverted Index</a></span></li></ul></li><li><span><a href="#5.5-Calculate-similarities---run-the-commands-and-insure-that-your-output-matches-the-output-below" data-toc-modified-id="5.5-Calculate-similarities---run-the-commands-and-insure-that-your-output-matches-the-output-below-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>5.5 Calculate similarities - run the commands and insure that your output matches the output below</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#NOTE:-you-must-run-in-hadoop-mode-to-generate-sorted-similarities" data-toc-modified-id="NOTE:-you-must-run-in-hadoop-mode-to-generate-sorted-similarities-8.4.0.1"><span class="toc-item-num">8.4.0.1&nbsp;&nbsp;</span>NOTE: you must run in hadoop mode to generate sorted similarities</a></span></li></ul></li><li><span><a href="#Pairwise-Similairity" data-toc-modified-id="Pairwise-Similairity-8.4.1"><span class="toc-item-num">8.4.1&nbsp;&nbsp;</span>Pairwise Similairity</a></span></li></ul></li></ul></li><li><span><a href="#===-END-OF-PHASE-1-===" data-toc-modified-id="===-END-OF-PHASE-1-===-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>=== END OF PHASE 1 ===</a></span></li></ul></div>

# MIDS - w261 Machine Learning At Scale
__Course Lead:__ Dr James G. Shanahan (__email__ Jimi via  James.Shanahan _AT_ gmail.com)

## Assignment - HW5


---
__Name:__  *Elizabeth Shulok*   
__Class:__ MIDS w261 Spring 2018 Group 2   
__Email:__  shulok@iSchool.Berkeley.edu     
__StudentId__  3032143336    __End of StudentId__     
__Week:__   5

__NOTE:__ please replace `1234567` with your student id above      
__Due Time:__ HW is due the Thursday of the following week by 8AM (West coast time). I.e., Thursday, Feb 14, 2017 in the case of this homework. 

* __HW5 Phase 1__ 
This can be done on a local machine (with a unit test on the cloud such as Altiscale's PaaS or on AWS) and is due Thursday, Week 6 by 8AM (West coast time). It will primarily focus on building a unit/systems and for pairwise similarity calculations pipeline (for stripe documents)

* __HW5 Phase 2__ 
This will require the Altiscale cluster and will be due Thursday, Feb 22 by 8AM (West coast time). 
The focus of  HW5 Phase 2  will be to scale up the unit/systems tests to the Google 5 gram corpus. 



### INSTRUCTIONS for SUBMISSIONS 
Follow the instructions for submissions carefully.

Each student has a `HW-<user>` repository for all assignments.   

Push the following to your HW github repo into the master branch:
* Your local HW5 directory. Your repo file structure should look like this:

```
HW-<user>
    --HW3
       |__MIDS-W261-HW-03-<Student_id>.ipynb
       |__MIDS-W261-HW-03-<Student_id>.pdf
       |__some other hw3 file
    --HW4
       |__MIDS-W261-HW-04-<Student_id>.ipynb
       |__MIDS-W261-HW-04-<Student_id>.pdf
       |__some other hw4 file
    etc..
```    

# Optional reading
http://stanford.edu/~rezab/papers/disco.pdf   
https://terpconnect.umd.edu/~oard/pdf/acl08elsayed2.pdf

# HW Problems

# HW5.0  data warehouse; star schema

- What is a data warehouse? What is a Star schema? When is it used?

> A data warehouse is a centralized repository that stores data from differen sources and is used for data analysis. A Star schema is the simpleset kind of data mart schema and it consists of fact tables and dimension tables. It is used when you have a data warehouse focused on a single subject and data is pulled from a limited number of sources. The data is not normalized and a Star schema is good for handling simple queries.

# HW5.1 Databases: 3NF; denormalized 

- In the database world What is 3NF? Does machine learning use data in 3NF? If so why? 
- In what form does ML consume data?
- Why would one use log files that are denormalized?

> 3NF is a normal form known as third normal form and is a common form for normalized data. Normalized data reduces redundant data storage an increases referential integroty. Machine learning can use data in 3NF, but often we use dimensionalized data.

> ML consumes data in both normalized form and dimensional. Often it uses a hybrid approach.

> Log files often have little structure. Using denormalized log files would allow for simpler quieries and faster aggregations.

# HW5.2  Memory-backed map-side

Using MRJob, implement a hashside join (memory-backed map-side) for left, right and inner joins. Use the following tables for this HW and join based on the country code (third column of the transactions table and the second column of the Countries table:

<pre>
transactions.dat
Alice Bob|$10|US
Sam Sneed|$1|CA
Jon Sneed|$20|CA
Arnold Wesise|$400|UK
Henry Bob|$2|US
Yo Yo Ma|$2|CA
Jon York|$44|CA
Alex Ball|$5|UK
Jim Davis|$66|JA

Countries.dat
United States|US
Canada|CA
United Kingdom|UK
Italy|IT

</pre>

**student task**


Please report the number of rows resulting from: (left = transactions, right = countries)

- (1) Left joining Table Left with Table Right
- (2) Right joining Table Left with Table Right
- (3) Inner joining Table Left with Table Right

__NOTE:__ For the Right and Inner joins, no reducer should be used. For the left outer join you can use a reducer.

__NOTE:__ If you need a refresher for MRJobs, please take a look at HW4.

> 1) 9 rows

> 2) 9 rows

> 3) 8 rows

In [1]:
%%writefile Countries.dat
United States|US
Canada|CA
United Kingdom|UK
Italy|IT

Writing Countries.dat


In [2]:
%%writefile transactions.dat
Alice Bob|$10|US
Sam Sneed|$1|CA
Jon Sneed|$20|CA
Arnold Wesise|$400|UK
Henry Bob|$2|US
Yo Yo Ma|$2|CA
Jon York|$44|CA
Alex Ball|$5|UK
Jim Davis|$66|JA

Writing transactions.dat


In [131]:
%%writefile performJoins.py

import sys, os, re
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRJoin(MRJob):
    ## mapper_init is responsible for reading in the countries table.
    def mapper_init(self):
        self.CCodes = {}
        self.UsedCodes = []
        with open("Countries.dat", "r") as infile:
            for line in infile:
                line = line.strip()
                data = line.split("|")
                if len(data) == 2:
                    country_code = data[1]
                    country = data[0]
                    self.CCodes[country_code] = country
        
class LeftJoin(MRJoin):
    
    def mapper(self, _, line):
        x = line.split("|")
        if len(x) == 3:
            name = x[0]
            amt = x[1]
            ccode = x[2]
            if self.CCodes.has_key(ccode):
                country = self.CCodes[ccode]
            else:
                country = "NULL"
            yield ccode, (name, amt, country)
        
class RightJoin(MRJoin):
    
    def steps(self):
        return [MRStep(
              mapper_init = self.mapper_init,
              mapper = self.mapper,
              reducer_init = self.reducer_init,
              reducer = self.reducer,
              reducer_final = self.reducer_final,
              jobconf = {'mapred.reduce.tasks': 1}
            )]

    def mapper(self, _, line):
        x = line.split("|")
        if len(x) == 3:
            name = x[0]
            amt = x[1]
            ccode = x[2]
            if self.CCodes.has_key(ccode):
                #if ccode not in self.UsedCodes:
                #    self.UsedCodes.append(ccode)
                country = self.CCodes[ccode]
                yield ccode, (name, amt, country)
    
    def reducer_init(self):
        self.CCodes = {}
        self.UsedCodes = []
        with open("Countries.dat", "r") as infile:
            for line in infile:
                line = line.strip()
                data = line.split("|")
                if len(data) == 2:
                    country_code = data[1]
                    country = data[0]
                    self.CCodes[country_code] = country
    
           
    # remove any duplicate lines by only yieliding unique records
    def reducer(self, key, values):
        if key not in self.UsedCodes:
            self.UsedCodes.append(key)
        for value in values:
            yield key, value
                
    def reducer_final(self):
        for key in self.CCodes:
            if key not in self.UsedCodes:
                yield key, ("NULL", self.CCodes[key])
            
class InnerJoin(MRJoin):
    
    def mapper(self, _, line):
        x = line.split("|")
        if len(x) == 3:
            name = x[0]
            amt = x[1]
            ccode = x[2]
            if self.CCodes.has_key(ccode):
                country = self.CCodes[ccode]
                yield ccode, (name, amt, country)
        
if __name__ == '__main__':
    print ("LeftJoin")
    LeftJoin.run()
    print ("\nRightJoin")
    RightJoin.run()
    print ("\nInnerJoin")
    InnerJoin.run()

Overwriting performJoins.py


In [132]:
!python performJoins.py transactions.dat --file Countries.dat > joins_output.csv

No configs found; falling back on auto-configuration
Creating temp directory /tmp/performJoins.root.20180214.053028.003670
Running step 1 of 1...
Streaming final output from /tmp/performJoins.root.20180214.053028.003670/output...
Removing temp directory /tmp/performJoins.root.20180214.053028.003670...
No configs found; falling back on auto-configuration
No configs found; falling back on auto-configuration
Creating temp directory /tmp/performJoins.root.20180214.053028.037001
Creating temp directory /tmp/performJoins.root.20180214.053028.037001
Running step 1 of 1...
Running step 1 of 1...
Streaming final output from /tmp/performJoins.root.20180214.053028.037001/output...
Streaming final output from /tmp/performJoins.root.20180214.053028.037001/output...
Removing temp directory /tmp/performJoins.root.20180214.053028.037001...
Removing temp directory /tmp/performJoins.root.20180214.053028.037001...
No configs found; falling back on auto-configuration
No configs found; falling back on auto

In [133]:
!cat joins_output.csv

LeftJoin
"US"	["Alice Bob","$10","United States"]
"CA"	["Sam Sneed","$1","Canada"]
"CA"	["Jon Sneed","$20","Canada"]
"UK"	["Arnold Wesise","$400","United Kingdom"]
"US"	["Henry Bob","$2","United States"]
"CA"	["Yo Yo Ma","$2","Canada"]
"CA"	["Jon York","$44","Canada"]
"UK"	["Alex Ball","$5","United Kingdom"]
"JA"	["Jim Davis","$66","NULL"]

RightJoin
"CA"	["Jon Sneed","$20","Canada"]
"CA"	["Jon York","$44","Canada"]
"CA"	["Sam Sneed","$1","Canada"]
"CA"	["Yo Yo Ma","$2","Canada"]
"UK"	["Alex Ball","$5","United Kingdom"]
"UK"	["Arnold Wesise","$400","United Kingdom"]
"US"	["Alice Bob","$10","United States"]
"US"	["Henry Bob","$2","United States"]
"IT"	["NULL","Italy"]

InnerJoin
"US"	["Alice Bob","$10","United States"]
"CA"	["Sam Sneed","$1","Canada"]
"CA"	["Jon Sneed","$20","Canada"]
"UK"	["Arnold Wesise","$400","United Kingdom"]
"US"	["Henry Bob","$2","United States"]
"CA"	["Yo Yo Ma","$2","Canada"]
"CA"	["Jon York","$44","Canada"]
"UK"	["Alex Ball","$5","United Kingdom"]


# HW5.2.1 (OPTIONAL) Almost stateless reducer-side join

The following MRJob code, implements a reduce-side join for an inner join. The reducer is almost stateless, i.e., uses as little memory as possible. Use the tables from HW5.2 for this HW and join based on the country code (third column of the transactions table and the second column of the Countries table perform. Perform  an left, right, inner joins using the code provided below and report the number of rows resulting from:

- (1) Left joining Table Left with Table Right
- (2) Right joining Table Left with Table Right
- (3) Inner joining Table Left with Table Right

Again make smart decisions about which table should be the left table (i.e., crosscheck the code). 

__Some notes on the code__ 
Here, the mapper receives its set of input splits either from the transaction table or from the countries table and makes the appropriate transformations: splitting the line into fields, and emitting a key/value. The key is the join key - in this case, the country code field of both sets of records. The mapper knows which file and type of record it is receiving based on the length of the fields. The records it emits contain the join field as the key, which acts as the partitioning key; We use the SORT_VALUES option, which ensures the values are sorted as well. Then, we employ a trick to ensure that for each join key, country records are seen always before transaction records. We achieve this by adding an arbitrary key to the front of the value: 'A' for countries, 'B' for customers. This makes countries sort before customers for each and every join/partition key. After that trick, the join is simply a matter of storing countries ('A' records) and crossing this array with each customer record.

In [ ]:
import sys, os, re
from mrjob.job import MRJob

class MRJoin(MRJob):

  # Performs secondary sort
  SORT_VALUES = True

  def mapper(self, _, line):
    splits = line.rstrip("\n").split("|")

    if len(splits) == 2: # country data
      symbol = 'A' # make country sort before transaction data
      country2digit = splits[1]
      yield country2digit, [symbol, splits]
    else: # person data
      symbol = 'B'
      country2digit = splits[2]
      yield country2digit, [symbol, splits]

  def reducer(self, key, values):
    countries = [] # should come first, as they are sorted on artificia key 'A'
    for value in values:
      if value[0] == 'A':
        countries.append(value)
      if value[0] == 'B':
        for country in countries:
          yield key, country[1:] + value[1:]

if __name__ == '__main__':
  MRJoin.run()

# Pairwise similarity  - PHASE 1

In this part of the assignment we will focus on developing methods for detecting synonyms, using the Google 5-grams dataset. To accomplish this you must script three main tasks using MRJob:


- (1) Using the systems tests data sets, write mrjob code to build the stripes
- (2) Write mrjob code to build an inverted index from the stripes
- (3) Using two (symmetric) comparison methods of your choice (e.g., correlations, distances, similarities), pairwise compare all stripes (vectors), and output to a file.   


For this task you will have to determine a method of comparison.
Here are a few that you might consider:

 - Jaccard
 - Cosine similarity
 - Spearman correlation
 - Euclidean distance
 - Taxicab (Manhattan) distance
 - Shortest path graph distance (a graph, because our data is symmetric!)
 - Pearson correlation
 - Kendall correlation
 
 ...
 
 (for more details: http://stanford.edu/~rezab/papers/disco.pdf pg5, table 1)

However, be cautioned that some comparison methods are more difficult to parallelize than others, and do not perform more associations than is necessary, since your choice of association will be symmetric.

Please use the inverted index (discussed in live session #5) based pattern to compute the pairwise (term-by-term) similarity matrix. 

## Run Systems tests locally on small datasets (PHASE1)

Complete 5.3-5.5 and systems test using the below test datasets. Phase 2 will focus on the entire Ngram dataset.

To help you through these tasks please verify that your code gives the results below (for stripes, inverted index, and pairwise similarities).

Test datasets:

* googlebooks-eng-all-5gram-20090715-0-filtered.txt [see below]
* atlas-boon-test [see below]
* stripe-docs-test [see below]


A large subset of the Google n-grams dataset   
https://aws.amazon.com/datasets/google-books-ngrams/

which we have placed in a bucket/folder on Dropbox:   
https://www.dropbox.com/sh/tmqpc4o0xswhkvz/AACUifrl6wrMrlK6a3X3lZ9Ea?dl=0 

You can download from Dropbox using wget:   
`wget -O googlebooks-eng-all-5gram-20090715-0-filtered.txt https://www.dropbox.com/sh/tmqpc4o0xswhkvz/AACr50woxiBWoaiiLmnwduX8a/googlebooks-eng-all-5gram-20090715-0-filtered.txt?dl=0`


In particular, this bucket contains (~200) files (10Meg each) in the format:

	(ngram) \t (count) \t (pages_count) \t (books_count)

The next cell shows the first 10 lines of the googlebooks-eng-all-5gram-20090715-0-filtered.txt file.


__DISCLAIMER__: Each record is already a 5-gram. In real life, we would calculate the stripes cooccurrence data from the raw text by windowing over the raw text and not from the 5-gram preprocessed data (as we are doing here).  Calculatating pairs on this 5-gram is a little corrupt as we will be double counting cooccurences. Having said that this exercise can still pull out some simialr terms. 

#### 1: unit/systems first-10-lines

In [1]:
%%writefile googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt
A BILL FOR ESTABLISHING RELIGIOUS	59	59	54
A Biography of General George	92	90	74
A Case Study in Government	102	102	78
A Case Study of Female	447	447	327
A Case Study of Limited	55	55	43
A Child's Christmas in Wales	1099	1061	866
A Circumstantial Narrative of the	62	62	50
A City by the Sea	62	60	49
A Collection of Fairy Tales	123	117	80
A Collection of Forms of	116	103	82

Overwriting googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt


#### 2: unit/systems atlas-boon

In [94]:
%%writefile atlas-boon-systems-test.txt
atlas boon	50	50	50
boon cava dipped	10	10	10
atlas dipped	15	15	15

Writing atlas-boon-systems-test.txt


#### 3: unit/systems stripe-docs-test

Three terms, A,B,C and their corresponding stripe-docs of co-occurring terms

- DocA {X:20, Y:30, Z:5}
- DocB {X:100, Y:20}
- DocC {M:5, N:20, Z:5}

## 5.3 build stripes for all the test data sets - run the commands and insure that your output matches the output below

* __Note__: Your results may not match 100%.

In [171]:
%%writefile buildStripes.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import re
import mrjob
import json
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRbuildStripes(MRJob):
  
    #START SUDENT CODE531_STRIPES
    
    # Performs secondary sort
    SORT_VALUES = True

    def mapper(self, _, line):
        x = line.rstrip("\n").split("\t")
        count = x[1]
        words = x[0].lower().split(" ")
        word_count = len(words)
        for n in range(word_count):
            for k in range(word_count):
                if n!=k:
                    yield words[n], (words[k], int(count))
  
    def reducer(self, key, values):
        currentList = {}
        # value is going to be word\tcount
        for value in values:
            word = value[0]
            count = value[1]
            if currentList.has_key(word):
                # if it's already in the list get the current count
                current_count = currentList[word]
                # add the new count
                currentList[word] = current_count + count
            else:
                currentList[word] = count
                
        yield key, currentList
  
    #END SUDENT CODE531_STRIPES
  
if __name__ == '__main__':
    MRbuildStripes.run()

Overwriting buildStripes.py


In [172]:
###########################################################################
# Make Stripes from ngrams for systems test 1
###########################################################################

!hdfs dfs rm --recursive systems_test_stripes_1
!python buildStripes.py -r local googlebooks-eng-all-5gram-20090715-0-filtered-first-10-lines.txt > systems_test_stripes_1

rm: Unknown command
Did you mean -rm?  This command begins with a dash.
No configs found; falling back on auto-configuration
Creating temp directory /tmp/buildStripes.root.20180214.062159.217124
Running step 1 of 1...
Streaming final output from /tmp/buildStripes.root.20180214.062159.217124/output...
Removing temp directory /tmp/buildStripes.root.20180214.062159.217124...


In [173]:
!cat systems_test_stripes_1

"a"	{"limited":55,"sea":62,"general":92,"female":447,"in":1201,"religious":59,"george":92,"biography":92,"city":62,"for":59,"tales":123,"government":102,"the":124,"forms":116,"wales":1099,"christmas":1099,"child's":1099,"collection":239,"by":62,"case":604,"circumstantial":62,"of":1011,"study":604,"bill":59,"establishing":59,"narrative":62,"fairy":123}
"bill"	{"a":59,"religious":59,"for":59,"establishing":59}
"biography"	{"a":92,"of":92,"george":92,"general":92}
"by"	{"a":62,"city":62,"the":62,"sea":62}
"case"	{"a":604,"limited":55,"government":102,"of":502,"study":604,"female":447,"in":102}
"child's"	{"a":1099,"wales":1099,"christmas":1099,"in":1099}
"christmas"	{"a":1099,"wales":1099,"in":1099,"child's":1099}
"circumstantial"	{"a":62,"of":62,"the":62,"narrative":62}
"city"	{"a":62,"the":62,"by":62,"sea":62}
"collection"	{"a":239,"forms":116,"fairy":123,"tales":123,"of":355}
"establishing"	{"a":59,"bill":59,"religious":59,"for":59}
"fairy"	{"a":123,"of":123,"tales":123,"collection":123

<pre>
"a"	{"limited": 55, "female": 447, "general": 92, "sea": 62, "in": 1201, "religious": 59, "george": 92, "biography": 92, "city": 62, "for": 59, "tales": 123, "child's": 1099, "by": 62, "forms": 116, "wales": 1099, "christmas": 1099, "government": 102, "collection": 239, "fairy": 123, "case": 604, "circumstantial": 62, "of": 1011, "study": 604, "bill": 59, "establishing": 59, "narrative": 62, "the": 124}
"bill"	{"a": 59, "religious": 59, "for": 59, "establishing": 59}
"biography"	{"a": 92, "of": 92, "george": 92, "general": 92}
"by"	{"a": 62, "city": 62, "the": 62, "sea": 62}
"case"	{"a": 604, "limited": 55, "government": 102, "of": 502, "study": 604, "female": 447, "in": 102}
"child's"	{"a": 1099, "wales": 1099, "christmas": 1099, "in": 1099}
"christmas"	{"a": 1099, "wales": 1099, "in": 1099, "child's": 1099}
"circumstantial"	{"a": 62, "of": 62, "the": 62, "narrative": 62}
"city"	{"a": 62, "the": 62, "by": 62, "sea": 62}
"collection"	{"a": 239, "forms": 116, "fairy": 123, "tales": 123, "of": 355}
"establishing"	{"a": 59, "bill": 59, "religious": 59, "for": 59}
"fairy"	{"a": 123, "of": 123, "tales": 123, "collection": 123}
"female"	{"a": 447, "case": 447, "study": 447, "of": 447}
"for"	{"a": 59, "bill": 59, "religious": 59, "establishing": 59}
"forms"	{"a": 116, "of": 232, "collection": 116}
"general"	{"a": 92, "of": 92, "george": 92, "biography": 92}
"george"	{"a": 92, "of": 92, "biography": 92, "general": 92}
"government"	{"a": 102, "case": 102, "study": 102, "in": 102}
"in"	{"a": 1201, "case": 102, "government": 102, "study": 102, "child's": 1099, "wales": 1099, "christmas": 1099}
"limited"	{"a": 55, "case": 55, "study": 55, "of": 55}
"narrative"	{"a": 62, "of": 62, "the": 62, "circumstantial": 62}
"of"	{"a": 1011, "case": 502, "circumstantial": 62, "george": 92, "limited": 55, "tales": 123, "collection": 355, "general": 92, "forms": 232, "female": 447, "narrative": 62, "study": 502, "fairy": 123, "the": 62, "biography": 92}
"religious"	{"a": 59, "bill": 59, "for": 59, "establishing": 59}
"sea"	{"a": 62, "city": 62, "the": 62, "by": 62}
"study"	{"a": 604, "case": 604, "limited": 55, "government": 102, "of": 502, "female": 447, "in": 102}
"tales"	{"a": 123, "of": 123, "fairy": 123, "collection": 123}
"the"	{"a": 124, "city": 62, "circumstantial": 62, "of": 62, "sea": 62, "narrative": 62, "by": 62}
"wales"	{"a": 1099, "in": 1099, "christmas": 1099, "child's": 1099}

</pre>

In [174]:
###########################################################################
# Make Stripes from ngrams for systems test 2
###########################################################################

!hdfs dfs rm --recursive systems_test_stripes_2
!python buildStripes.py -r local atlas-boon-systems-test.txt > systems_test_stripes_2

rm: Unknown command
Did you mean -rm?  This command begins with a dash.
No configs found; falling back on auto-configuration
Creating temp directory /tmp/buildStripes.root.20180214.062214.452397
Running step 1 of 1...
Streaming final output from /tmp/buildStripes.root.20180214.062214.452397/output...
Removing temp directory /tmp/buildStripes.root.20180214.062214.452397...


In [175]:
!cat systems_test_stripes_2

"atlas"	{"dipped":15,"boon":50}
"boon"	{"atlas":50,"dipped":10,"cava":10}
"cava"	{"dipped":10,"boon":10}
"dipped"	{"atlas":15,"boon":10,"cava":10}


<pre>
"atlas"   {"dipped": 15, "boon": 50}   
"boon"    {"atlas": 50, "dipped": 10, "cava": 10}   
"cava"    {"dipped": 10, "boon": 10} 
"dipped"  {"atlas": 15, "boon": 10, "cava": 10}
</pre>

In [176]:
########################################################################
# Stripes for systems test 3 (given, no need to build stripes)
########################################################################

with open("systems_test_stripes_3", "w") as f:
    f.writelines([
        '"DocA"\t{"X":20, "Y":30, "Z":5}\n',
        '"DocB"\t{"X":100, "Y":20}\n',  
        '"DocC"\t{"M":5, "N":20, "Z":5, "Y":1}\n'
    ])
!cat systems_test_stripes_3   

"DocA"	{"X":20, "Y":30, "Z":5}
"DocB"	{"X":100, "Y":20}
"DocC"	{"M":5, "N":20, "Z":5, "Y":1}


## 5.4 Build Inverted Index - run the commands and insure that your output matches the output below

* You might find this [paper](https://terpconnect.umd.edu/~oard/pdf/acl08elsayed2.pdf) useful.

In [192]:
%%writefile invertedIndex.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-


from __future__ import division
import collections
import re
import json
import math
import numpy as np
import itertools
import mrjob
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRinvertedIndex(MRJob):
    
    #START SUDENT CODE531_INV_INDEX

    # Performs secondary sort
    SORT_VALUES = True
    
    def mapper(self, _, line):
        line = line.strip()
        word,stripe = line.split("\t")
        stripe = json.loads(stripe)
        vector_length = len(stripe)
        for key in stripe:
            yield key, (word.replace('"', ''), vector_length)
        
    def reducer(self, key, values):
        currentList = []
        # value is going to be word\tlist
        for value in values:
            word = value[0]
            list = value[1]
            if currentList.has_key(word):
                currentList[word] = currentList[word].append(list)
            else:
                currentList[word] = list
                
        yield key, currentList
    
    #END SUDENT CODE531_INV_INDEX
        
if __name__ == '__main__':
    MRinvertedIndex.run() 

Overwriting invertedIndex.py


In [193]:
!python invertedIndex.py -r local systems_test_stripes_1 > systems_test_index_1

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndex.root.20180214.072758.158898
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20180214.072758.158898/output...
Removing temp directory /tmp/invertedIndex.root.20180214.072758.158898...


In [194]:
!python invertedIndex.py -r local systems_test_stripes_2 > systems_test_index_2

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndex.root.20180214.072840.935534
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20180214.072840.935534/output...
Removing temp directory /tmp/invertedIndex.root.20180214.072840.935534...


In [195]:
!python invertedIndex.py -r local systems_test_stripes_3 > systems_test_index_3

No configs found; falling back on auto-configuration
Creating temp directory /tmp/invertedIndex.root.20180214.072844.733534
Running step 1 of 1...
Streaming final output from /tmp/invertedIndex.root.20180214.072844.733534/output...
Removing temp directory /tmp/invertedIndex.root.20180214.072844.733534...


In [196]:
##########################################################
# Pretty print systems tests for generating Inverted Index
##########################################################

import json

for i in range(1,4):
    print "—"*100
    print "Systems test ",i," - Inverted Index"
    print "—"*100  
    with open("systems_test_index_"+str(i),"r") as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            word,stripe = line.split("\t")
            stripe = json.loads(stripe)
            stripe.extend([["",""] for _ in xrange(3 - len(stripe))])

            print "{0:>16} |{1:>16} |{2:>16} |{3:>16}".format(
              (word), stripe[0][0]+" "+str(stripe[0][1]), stripe[1][0]+" "+str(stripe[1][1]), stripe[2][0]+" "+str(stripe[2][1]))



————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |          bill 4 |     biography 4 |            by 4
          "bill" |            a 27 |  establishing 4 |           for 4
     "biography" |            a 27 |       general 4 |        george 4
            "by" |            a 27 |          city 4 |           sea 4
          "case" |            a 27 |        female 4 |    government 4
       "child's" |            a 27 |     christmas 4 |            in 7
     "christmas" |            a 27 |       child's 4 |            in 7
"circumstantial" |            a 27 |     narrative 4 |           of 16
          "city" |            a 27 |            by 4 |           sea 4
    "collection" |            a 27 |         fairy 4 |         forms 3
  "establishing" |            a 27 |          bill 4 |

### Inverted Index

In [ ]:
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "a" |          bill 4 |     biography 4 |            by 4
          "bill" |            a 27 |  establishing 4 |           for 4
     "biography" |            a 27 |       general 4 |        george 4
            "by" |            a 27 |          city 4 |           sea 4
          "case" |            a 27 |        female 4 |    government 4
       "child's" |            a 27 |     christmas 4 |            in 7
     "christmas" |            a 27 |       child's 4 |            in 7
"circumstantial" |            a 27 |     narrative 4 |           of 15
          "city" |            a 27 |            by 4 |           sea 4
    "collection" |            a 27 |         fairy 4 |         forms 3
  "establishing" |            a 27 |          bill 4 |           for 4
         "fairy" |            a 27 |    collection 5 |           of 15
        "female" |            a 27 |          case 7 |           of 15
           "for" |            a 27 |          bill 4 |  establishing 4
         "forms" |            a 27 |    collection 5 |           of 15
       "general" |            a 27 |     biography 4 |        george 4
        "george" |            a 27 |     biography 4 |       general 4
    "government" |            a 27 |          case 7 |            in 7
            "in" |            a 27 |          case 7 |       child's 4
       "limited" |            a 27 |          case 7 |           of 15
     "narrative" |            a 27 |circumstantial 4 |           of 15
            "of" |            a 27 |     biography 4 |          case 7
     "religious" |            a 27 |          bill 4 |  establishing 4
           "sea" |            a 27 |            by 4 |          city 4
         "study" |            a 27 |          case 7 |        female 4
         "tales" |            a 27 |    collection 5 |         fairy 4
           "the" |            a 27 |            by 4 |circumstantial 4
         "wales" |            a 27 |       child's 4 |     christmas 4
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  2  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
         "atlas" |          boon 3 |        dipped 3 |                
          "boon" |         atlas 2 |          cava 2 |        dipped 3
          "cava" |          boon 3 |        dipped 3 |                
        "dipped" |         atlas 2 |          boon 3 |          cava 2
————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  3  - Inverted Index
————————————————————————————————————————————————————————————————————————————————————————————————————
             "M" |          DocC 4 |                 |                
             "N" |          DocC 4 |                 |                
             "X" |          DocA 3 |          DocB 2 |                
             "Y" |          DocA 3 |          DocB 2 |          DocC 4
             "Z" |          DocA 3 |          DocC 4 |                


## 5.5 Calculate similarities - run the commands and insure that your output matches the output below

In [120]:
%%writefile similarity.py
#!~/anaconda2/bin/python
# -*- coding: utf-8 -*-

from __future__ import division
import collections
import re
import json
import math
#import numpy as np
import itertools
import mrjob
from mrjob.protocol import RawProtocol
from mrjob.job import MRJob
from mrjob.step import MRStep

class MRsimilarity(MRJob):
  
    #START SUDENT CODE531_SIMILARITY

    # Performs secondary sort
    SORT_VALUES = True
    
    # Jaccard Similarity = |A ∩ B| : |A ∪ B|
    
    def mapper(self, _, line):
        line = line.strip()
        word,stripe = line.split("\t")
        words = json.loads(stripe)
        vector_length = len(words)
        for n in range(vector_length):
            for k in range(n+1,vector_length):
                yield words[n][0] + "-" + words[k][0], (words[n][1], words[k][1])
    
    def reducer(self, key, values):
        # value is going to be word\tlist
        sum = 0
        count = 0
        x = y = 0
        for value in values:
            count = count + 1
            x = value[0]
            y = value[1]
        jaccard = count/(x+y-count)
        cosine = count/(math.sqrt(x)*math.sqrt(y))
        average = (jaccard+cosine)/2
        
        yield average, (key, cosine, jaccard)
            
           
    #END SUDENT CODE531_SIMILARITY
  
if __name__ == '__main__':
    MRsimilarity.run()

Overwriting similarity.py


#### NOTE: you must run in hadoop mode to generate sorted similarities

In [121]:
!python similarity.py -r hadoop systems_test_index_1 > systems_test_similarities_1

No configs found; falling back on auto-configuration
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20180217.155119.605952
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20180217.155119.605952/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob4753382548296440126.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1518738241717_0010
  Submitted application application_1518738241717_0010
  The url to track the job: http://quickstart.cloudera:8088/proxy/applicati

In [122]:
!python similarity.py -r hadoop systems_test_index_2 > systems_test_similarities_2

No configs found; falling back on auto-configuration
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20180217.155213.346779
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20180217.155213.346779/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob3298856988533390386.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1518738241717_0011
  Submitted application application_1518738241717_0011
  The url to track the job: http://quickstart.cloudera:8088/proxy/applicati

In [123]:
!python similarity.py -r hadoop systems_test_index_3 > systems_test_similarities_3

No configs found; falling back on auto-configuration
Looking for hadoop binary in $PATH...
Found hadoop binary: /usr/bin/hadoop
Using Hadoop version 2.6.0
Looking for Hadoop streaming jar in /home/hadoop/contrib...
Looking for Hadoop streaming jar in /usr/lib/hadoop-mapreduce...
Found Hadoop streaming jar: /usr/lib/hadoop-mapreduce/hadoop-streaming.jar
Creating temp directory /tmp/similarity.root.20180217.155404.376571
Copying local files to hdfs:///user/root/tmp/mrjob/similarity.root.20180217.155404.376571/files/...
Running step 1 of 1...
  packageJobJar: [] [/usr/jars/hadoop-streaming-2.6.0-cdh5.7.0.jar] /tmp/streamjob3414215400098885457.jar tmpDir=null
  Connecting to ResourceManager at /0.0.0.0:8032
  Connecting to ResourceManager at /0.0.0.0:8032
  Total input paths to process : 1
  number of splits:2
  Submitting tokens for job: job_1518738241717_0012
  Submitted application application_1518738241717_0012
  The url to track the job: http://quickstart.cloudera:8088/proxy/applicati

In [125]:
############################################
# Pretty print systems tests
# Note: adjust print formatting if you need to
############################################

import json
for i in range(1,4):
    print '—'*110
    print "Systems test ",i," - Similarity measures"
    print '—'*110
    #print "{0:>15} |{1:>15} |{2:>15} |{3:>15} |{4:>15} |{5:>15}".format(
    #        "average", "pair", "cosine", "jaccard", "overlap", "dice")

    print "{0:>15} |{1:>15} |{2:>15} |{3:>15}".format(
            "average", "pair", "cosine", "jaccard")

    print '-'*110

    with open("systems_test_similarities_"+str(i),"r") as f:
        lines = f.readlines()
        for line in lines:
            line = line.strip()
            avg,stripe = line.split("\t")
            stripe = json.loads(stripe)

            #print "{0:>15f} |{1:>15} |{2:>15f} |{3:>15f} |{4:>15f} |{5:>15f}".format(
            #  float(avg), stripe[0], float(stripe[1]), float(stripe[2]), float(stripe[3]), float(stripe[4]))

            print "{0:>15f} |{1:>15} |{2:>15f} |{3:>15}".format(
                float(avg), stripe[0], float(stripe[1]), float(stripe[2]))

——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
        average |           pair |         cosine |        jaccard
--------------------------------------------------------------------------------------------------------------
       0.197909 |         a-bill |       0.288675 | 0.107142857143
       0.197909 |    a-biography |       0.288675 | 0.107142857143
       0.197909 |           a-by |       0.288675 | 0.107142857143
       0.325361 |         a-case |       0.436436 | 0.214285714286
       0.197909 |      a-child's |       0.288675 | 0.107142857143
       0.197909 |    a-christmas |       0.288675 | 0.107142857143
       0.197909 |a-circumstantial |       0.288675 | 0.107142857143
       0.197909 |         a-city |       0.288675 | 0.107142857143
       0.243561 |   a-co

### Pairwise Similairity 

In [ ]:
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  1  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  1.000000 |    female - limited |       1.000000 |       1.000000 |       1.000000 |       1.000000
  0.868292 |       fairy - forms |       0.866025 |       0.750000 |       1.000000 |       0.857143
  0.868292 |       forms - tales |       0.866025 |       0.750000 |       1.000000 |       0.857143
  0.830357 |        case - study |       0.857143 |       0.750000 |       0.857143 |       0.857143
  0.712500 | bill - establishing |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |   christmas - wales |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |circumstantial - narrative |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |            by - sea |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |           by - city |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |     child's - wales |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 |  biography - george |       0.750000 |       0.600000 |       0.750000 |       0.750000
  0.712500 | child's - christmas |       0.750000 |       0.600000 |       0.750000 |       0.750000
  ...
  
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  2  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  1.000000 |        atlas - cava |       1.000000 |       1.000000 |       1.000000 |       1.000000
  0.625000 |       boon - dipped |       0.666667 |       0.500000 |       0.666667 |       0.666667
  0.389562 |       cava - dipped |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |         boon - cava |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |      atlas - dipped |       0.408248 |       0.250000 |       0.500000 |       0.400000
  0.389562 |        atlas - boon |       0.408248 |       0.250000 |       0.500000 |       0.400000
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
Systems test  3  - Similarity measures
——————————————————————————————————————————————————————————————————————————————————————————————————————————————
   average |                pair |         cosine |        jaccard |        overlap |           dice
--------------------------------------------------------------------------------------------------------------
  0.820791 |         DocA - DocB |       0.816497 |       0.666667 |       1.000000 |       0.800000
  0.553861 |         DocA - DocC |       0.577350 |       0.400000 |       0.666667 |       0.571429
  0.346722 |         DocB - DocC |       0.353553 |       0.200000 |       0.500000 |       0.333333


# === END OF PHASE 1 ===